### PR TITLE
Rename lsif-server to precise-code-intel

### PR DESCRIPTION
**Pull request by efritz**
This is the pure docker/docker compose analog for https://github.com/sourcegraph/deploy-sourcegraph/pull/574.

Removes the lsif-server containers in favor of precise-code-intel-api-server, precise-code-intel-bundle-manager, and precise-code-intel-worker.

### CODE REVIEW

File: prometheus/prometheus_targets.yml [hunk 1]
```diff
@@ -12,8 +12,8 @@
     - query-runner:6060
     - repo-updater:6060
     - syntect-server:6060
-    - lsif-server:3186
-    - lsif-server:3187
+    - precise-code-intel-api-server:3186
+    - precise-code-intel-bundle-manager:3187
```
Comment (slimsag) [inline prometheus/prometheus_targets.yml:16]: Where is the entry or the worker?

File: docker-compose/docker-compose.yaml [hunk 2]
```diff
@@ -252,29 +246,76 @@ services:
...
+  # Ports exposed to other Sourcegraph services: 3188/TCP (first worker), 9090 (prometheus)
+  # Ports exposed to the public internet: none
+  #
+  precise-code-intel-worker:
+    container_name: precise-code-intel-worker
+    image: 'index.docker.io/sourcegraph/precise-code-intel-worker:insiders'
+    cpus: 2
+    mem_limit: '4g'
+    environment:
+      - NUM_WORKERS=4
```
Comment (slimsag) [inline docker-compose/docker-compose.yaml:310]: Can we rely on the default (2) value here instead?

File: prometheus/prometheus_targets.yml (hunk 1 above)
Comment (efritz) [reply prometheus/prometheus_targets.yml:16]: It's federated in prometheus as it can run more than one metrics endpoint.

File: docker-compose/docker-compose.yaml (hunk 2 above)
Comment (efritz) [reply docker-compose/docker-compose.yaml:310]: No problem.

File: docker-compose/docker-compose.yaml [hunk 3]
```diff
@@ -252,29 +246,76 @@ services:
...
+  #
+  precise-code-intel-bundle-manager:
+    container_name: precise-code-intel-bundle-manager
+    image: 'index.docker.io/sourcegraph/precise-code-intel-bundle-manager:insiders'
+    cpus: 2
+    mem_limit: '2g'
+    environment:
+      - PRECISE_CODE_INTEL_API_SERVER_URL=http://precise-code-intel-api-server:3186
+      - 'SRC_FRONTEND_INTERNAL=sourcegraph-frontend-internal:3090'
+      - LSIF_STORAGE_ROOT=/lsif-storage
```
Comment (slimsag) [inline docker-compose/docker-compose.yaml:286]: Why is this here?

File: deploy-precise-code-intel-bundle-manager.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+#
+docker run --detach \
+    --name=precise-code-intel-bundle-manager \
+    --network=sourcegraph \
+    --restart=always \
+    --cpus=2 \
+    --memory=2g \
+    -e 'PRECISE_CODE_INTEL_API_SERVER_URL=http://precise-code-intel-api-server:3186' \
+    -e 'SRC_FRONTEND_INTERNAL=sourcegraph-frontend-internal:3090' \
+    -e 'LSIF_STORAGE_ROOT=/lsif-storage' \
```
Comment (slimsag) [inline deploy-precise-code-intel-bundle-manager.sh:19]: shouldn't be here

File: deploy-precise-code-intel-worker.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+# Ports exposed to other Sourcegraph services: 3188/TCP (first worker), 9090 (prometheus)
+# Ports exposed to the public internet: none
+#
+docker run --detach \
+    --name=precise-code-intel-worker \
+    --network=sourcegraph \
+    --restart=always \
+    --cpus=2 \
+    --memory=4g \
+    -e 'NUM_WORKERS=2' \
```
Comment (slimsag) [inline deploy-precise-code-intel-worker.sh:16]: shouldn't be here, we should be able to rely on the default being 2

File: docker-compose/docker-compose.yaml (hunk 3 above)
Comment (efritz) [reply docker-compose/docker-compose.yaml:286]: The bundle manager still has a volume mounted at /lsif-storage. We can remove this as it's the default value for that envvar, but 294 would need to stay the same.

File: deploy-precise-code-intel-worker.sh (hunk 5 above)
Comment (efritz) [reply deploy-precise-code-intel-worker.sh:16]: Do you mean the default in the docker container?

File: deploy-precise-code-intel-bundle-manager.sh (hunk 4 above)
Comment (efritz) [reply deploy-precise-code-intel-bundle-manager.sh:19]: See above.

File: deploy-precise-code-intel-worker.sh (hunk 5 above)
Comment (slimsag) [reply deploy-precise-code-intel-worker.sh]: Yes

Comment (efritz) [reply deploy-precise-code-intel-worker.sh]: I'm removing all NUM_WORKERS from this repo and will change the image default to 2 in sg/sg.

File: docker-compose/docker-compose.yaml (hunk 3 above)
Comment (slimsag) [reply docker-compose/docker-compose.yaml:286]: Makes sense, yes we should rely on the default value.

File: deploy-precise-code-intel-bundle-manager.sh (hunk 4 above)
Comment (slimsag) [reply deploy-precise-code-intel-bundle-manager.sh:19]: We should rely on the default value.

File: docker-compose/docker-compose.yaml [hunk 6]
```diff
@@ -252,29 +246,75 @@ services:
...
+
+  # Description: Stores and manages precise code intelligence bundles.
+  #
+  # Disk: 200GB / persistent SSD
+  # Ports exposed to other Sourcegraph services: 3187/TCP (manager)
+  # Ports exposed to the public internet: none
+  #
+  precise-code-intel-bundle-manager:
+    container_name: precise-code-intel-bundle-manager
+    image: 'index.docker.io/sourcegraph/precise-code-intel-bundle-manager:insiders'
```
Comment (slimsag) [inline docker-compose/docker-compose.yaml:280]: It's OK to merge this with `insiders` tags, but they need to be pinned using SHA digests.

File: docker-compose/docker-compose.yaml [hunk 7]
```diff
@@ -252,29 +246,75 @@ services:
...
+  # Ports exposed to other Sourcegraph services: 3188/TCP (first worker), 9090 (prometheus)
+  # Ports exposed to the public internet: none
+  #
+  precise-code-intel-worker:
+    container_name: precise-code-intel-worker
+    image: 'index.docker.io/sourcegraph/precise-code-intel-worker:insiders'
+    cpus: 2
+    mem_limit: '4g'
+    environment:
+      - PRECISE_CODE_INTEL_BUNDLE_MANAGER_URL=http://precise-code-intel-bundle-manager:3187
```
Comment (slimsag) [inline docker-compose/docker-compose.yaml:310]: For consistency, the env vars you added to this file should also be in single quotes like all others in this file

Comment (efritz) [reply docker-compose/docker-compose.yaml]: Done.

File: docker-compose/docker-compose.yaml (hunk 6 above)
Comment (efritz) [reply docker-compose/docker-compose.yaml]: Updated to same commit as the others.

File: deploy-precise-code-intel-bundle-manager.sh (hunk 4 above)
Comment (efritz) [reply deploy-precise-code-intel-bundle-manager.sh]: Removed.

File: prometheus/prometheus.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  - job_name: "sg"
+    file_sd_configs:
+      - files:
+          - "/sg_prometheus_add_ons/*_targets.yml"
+
+## Note: the content above is taken from sourcegraph/sourcegraph/docker-images/prometheus/config/prometheus.yml
+## I'm not currently sure how to configure the below in an external file via file_sd_configs.
+##
+## TODO(eric,uwedeportivo) - find a way to express what's below in an extensible way so we can 
+## just add config files in this directory instead of overwriting the entire main configuration.
```
Comment (slimsag) [inline prometheus/prometheus.yml:36]: We can't merge this PR as-is with this in place. We must move towards less configuration in this repository and not more, introducing a new config file here is IMO not an option.

A few options I would be content with:

1. Merge this PR without monitoring for the workers, and defer this work to another PR. I am fine with that, but we would need to figure it out before 3.15 releases.
2. An environment variable on the prometheus image which controls whether or not this configuration is loaded (not a file stored in this repo).

Comment (efritz) [reply prometheus/prometheus.yml:36]: We can remove monitoring, but I'm not 100% we can fix this before the 3.15 release in the way @sourcegraph/code-intel wants (we're trying to rewrite parts of the worker in Go so we don't have this problem which is caused by running multiple node processes in the same container).

cc @uwedeportivo

Comment (slimsag) [reply prometheus/prometheus.yml:36]: I don't think it would be too hard to solve, though?

Right now in `master` we are using the builtin `prometheus.yml` here: https://github.com/sourcegraph/sourcegraph/blob/master/docker-images/prometheus/config/prometheus.yml

You would just need to copy this prometheus.yml into the same location, but name it say e.g. `prometheus_pure_docker.yml` and then use it by setting the env var here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@6fef3d27e4938dd8d750afd7a81dd6a150d04e9b/-/blob/docker-images/prometheus/entry.sh#L4-8

That wouldn't be perfect (because you would still have `precise-code-intel-worker:9090` hard-coded in that file) but that'd be good enough for 3.15.

Then in e.g. `deploy-prometheus.sh` in this repo you would just add `-e PURE_DOCKER=true` to trigger using that config file.

Comment (efritz) [reply prometheus/prometheus.yml:36]: Ok that seems like a fine stopgap.

Comment (efritz) [reply prometheus/prometheus.yml:36]: Working on this now. And for context, this entire federation technique should be able to be removed in a few cycles once we rewrite some of the worker wrapper in Go and can use the language concurrency rather than the OS's.

Comment (efritz) [reply prometheus/prometheus.yml]: See https://github.com/sourcegraph/sourcegraph/pull/9552.

I'm unsure of how to update the prometheus image, though.
